core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -488,18 +489,21 @@ class LogSegment(val log: FileRecords,
    * Delete this log segment from the filesystem.
    */
   def delete() {
-    val deletedLog = log.delete()
-    val deletedIndex = index.delete()
-    val deletedTimeIndex = timeIndex.delete()
-    val deletedTxnIndex = txnIndex.delete()
-    if (!deletedLog && log.file.exists)
-      throw new IOException("Delete of log " + log.file.getName + " failed.")
-    if (!deletedIndex && index.file.exists)
-      throw new IOException("Delete of index " + index.file.getName + " failed.")
-    if (!deletedTimeIndex && timeIndex.file.exists)
-      throw new IOException("Delete of time index " + timeIndex.file.getName + " failed.")
-    if (!deletedTxnIndex && txnIndex.file.exists)
-      throw new IOException("Delete of transaction index " + txnIndex.file.getName + " failed.")
+    def closeable(delete: () => Unit, fileType: String, file: File): Closeable = new Closeable {
+      def close(): Unit =
+        try delete()
+        catch {
+          case e: IOException => throw new IOException(s"Delete of $fileType ${file.getName} failed.", e)
+        }
+    }
+
+    val closeables = Seq(
+      closeable(log.delete _, "log", log.file),
+      closeable(index.delete _, "index", index.file),
+      closeable(timeIndex.delete _, "time index", timeIndex.file),
+      closeable(txnIndex.delete _, "transaction index", txnIndex.file),
```

Comment:
remove the , at the end?

Reply:
Nice catch. This was disallowed at some point and then it became allowed again (sadly) after many people asked for it. I compiled with 2.12.4 and probably why it didn't complain.